### DISCUSSION TITLE
write displacement field

### DISCUSSION THREAD

**Discussion by XuanBaiBai** [andreasmang/claire#12] [Q&A]
After using "-deffield" to write three scalar displacement fields, what should I do to merge them to obtain a vector field which can be displayed in ITK-snap, just like the results of  GetDisplacementField in SimpleITK? 

like this picture...
![1673270343446](https://user-images.githubusercontent.com/33119184/211317179-6d6bf5c6-9781-4c98-bba8-dd43a91cd9f6.png)

Looking forward to reply, thanks a lot.

**Comment by andreasmang** [comment]
Dear @XuanBaiBai: 

Thanks for reaching out.

Generally speaking, we "apply" the diffeomorphism by solving the transport problem, i.e., use the computed velocity and solve the forward problem for a given image. So, to generate something similar to the image you show above, you would create a template image that overlays a regular mesh to the template image by coloring some of the voxels brightly.

If you want to visualize the computed deformation map (diffeomorphism) the way we do it in our publications: What we do is load the deformation field and plot it in Matlab and overlay it on the deformed template image (again done in Matlab; of course this can be written in any language that is good at plotting / visualization).

https://github.com/andreasmang/claire/blob/gpu/src/CLAIREBase.cpp#L2006
https://github.com/andreasmang/claire/blob/gpu/src/DeformationFields/DeformationFields.cpp#L1195
https://github.com/andreasmang/claire/blob/gpu/src/DeformationFields/DeformationFields.cpp#L1195

That is, we solve the flow equation d_t y = v(y) to compute the deformation map / diffeomorphism y inside CLAIRE. CLAIRE provides this as an output of the solver ("-defmap" option): 

https://github.com/andreasmang/claire/blob/gpu/src/RegOpt.cpp#L606

Notice that applying the velocity to an image (like I explained in the beginning) is the preferred method to deform images (and not go through the deformation map (i.e., load the image and apply the diffeomorphism by interpolation). The reason is that the map does not appear in our formulation; the transport equation does. So, you more accurately reflect what the optimizer "sees" in your output.

I am happy to share code for the visualization if you need it. May take some time to get it into shape. It's inside some Matlab toolbox we developed for prototyping. 

I hope this answers your question (at least partially). I can add additional material / look into things if there are any issues.
